pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,621 @@
+    : client_(client),
+      subscriptionName_(subscriptionName),
+      topic_(topicName ? topicName->toString() : "EmptyTopics"),
+      conf_(conf),
+      state_(Pending),
+      messages_(1000),
+      listenerExecutor_(client->getListenerExecutorProvider()->get()),
+      messageListener_(conf.getMessageListener()),
+      namespaceName_(topicName ? topicName->getNamespaceName() : boost::shared_ptr<NamespaceName>()),
+      lookupServicePtr_(lookupServicePtr),
+      allTopicPartitionsNumber_(boost::make_shared<std::atomic<int>>(0)),
+      topics_(topics) {
+    std::stringstream consumerStrStream;
+    consumerStrStream << "[Muti Topics Consumer: "
+                      << "TopicName - " << topic_ << " - Subscription - " << subscriptionName << "]";
+    consumerStr_ = consumerStrStream.str();
+
+    if (conf.getUnAckedMessagesTimeoutMs() != 0) {
+        unAckedMessageTrackerPtr_.reset(
+            new UnAckedMessageTrackerEnabled(conf.getUnAckedMessagesTimeoutMs(), client, *this));
+    } else {
+        unAckedMessageTrackerPtr_.reset(new UnAckedMessageTrackerDisabled());
+    }
+}
+
+void MultiTopicsConsumerImpl::start() {
+    if (topics_.empty()) {
+        setState(Ready);
+        LOG_DEBUG("No topics passed in when create MultiTopicsConsumer.");
+        multiTopicsConsumerCreatedPromise_.setValue(shared_from_this());
+        return;
+    }
+
+    // start call subscribeOneTopicAsync for each single topic
+    int topicsNumber = topics_.size();
+    boost::shared_ptr<std::atomic<int>> topicsNeedCreate = boost::make_shared<std::atomic<int>>(topicsNumber);
+    // subscribe for each passed in topic
+    for (std::vector<std::string>::const_iterator itr = topics_.begin(); itr != topics_.end(); itr++) {
+        subscribeOneTopicAsync(*itr).addListener(
+            boost::bind(&MultiTopicsConsumerImpl::handleOneTopicSubscribed, shared_from_this(), _1, _2, *itr,
+                        topicsNeedCreate));
+    }
+}
+
+void MultiTopicsConsumerImpl::handleOneTopicSubscribed(Result result, Consumer consumer,
+                                                       const std::string& topic,
+                                                       boost::shared_ptr<std::atomic<int>> topicsNeedCreate) {
+    int previous = topicsNeedCreate->fetch_sub(1);
+    assert(previous > 0);
+
+    if (result != ResultOk) {
+        state_ = Failed;
+        multiTopicsConsumerCreatedPromise_.setFailed(result);
+        // unsubscribed all of the successfully subscribed partitioned consumers
+        ResultCallback nullCallbackForCleanup = NULL;
+        closeAsync(nullCallbackForCleanup);
+        LOG_ERROR("Unable to create Consumer - "
+                  << " Error - " << result);
+        return;
+    }
+
+    LOG_DEBUG("Successfully Subscribed to topic " << topic << " in TopicsConsumer ");
+
+    if (topicsNeedCreate->load() == 0) {
+        LOG_INFO("Successfully Subscribed to Topics");
+        setState(Ready);
```

Comment:
What if a previous topic in the list had failed to subscribe? The state would have been set to Failed, but then it's getting set to Ready here.